### ISSUE TITLE
Not OK: perl 5.11.5 on ia64-linux-ld 2.6.29.2 (UNINSTALLED)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10224]
Migrated from [rt.perl.org#73496](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=73496) (status was 'resolved')

Searchable as RT73496$

**Comment by p5pRT** [comment]
### From @nwc10



\.\./lib/overload\.t \.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\. 1/607
\#   Failed test 'numifies to usual reference value'
\#   at \.\./lib/overload\.t line 1561\.
\#          got&#8203;: '6\.91752902764528267e\+18'
\#     expected&#8203;: '6917529027645282672'

\#   Failed test 'division of ref'
\#   at \.\./lib/overload\.t line 1572\.
\#          got&#8203;: '6\.91752902764528267e\+18'
\#     expected&#8203;: '6917529027645282672'
\# Looks like you failed 2 tests of 607\.
\.\./lib/overload\.t \.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\. Dubious\, test returned 2 \(wstat 512\, 0x200\)
Failed 2/607 subtests

The tests in question are

&nbsp;    is\($m/1\, $num\_val\, 'numifies to usual reference value'\);

&nbsp;    is\($aref/1\, $num\_val\, 'division of ref'\);

where $m and $aref are created like this&#8203;:

&nbsp;    package numify\_by\_fallback;
&nbsp;    use overload fallback => 1;

&nbsp;    package main;

&nbsp;    my $aref = \[\];
&nbsp;    my $m = bless $aref\, 'numify\_by\_fallback';


The test is making unwarrented assumptions about the numeric values of the
memory map\. On ia64 Linux the values are BIG&#8203;:

$ \./perl \-e 'printf "%x\\n"\, \[\]'
600000000001f0e0

I'm not sure what the right fix is\. I don't think that this constitutes a
blocker\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=install
    severity=none

Site configuration information for perl 5.11.5:

Configured by nick at Wed Mar 10 19:27:50 CET 2010.

Summary of my perl5 (revision 5 version 11 subversion 5) configuration:
  Derived from: ab8e66c141d8d7d892995deba69405e747204cf0
  Platform:
    osname=linux, osvers=2.6.29.2, archname=ia64-linux-ld
    uname='linux babe 2.6.29.2 #1 smp wed apr 29 16:24:15 cest 2009 ia64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-O -Uusethreads -Uuse64bitall -Duselongdouble -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.11.5-89-gab8e66c -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.11.5-89-gab8e66c -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O',
    cppflags='-fno-strict-aliasing -pipe'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='long double', nvsize=16, Off_t='off_t', lseeksize=8
    alignbytes=16, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.11.5:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.11.5-89-gab8e66c/lib/perl5/site_perl/5.11.5/ia64-linux-ld
    /home/nick/Sandpit/snap5.9.x-v5.11.5-89-gab8e66c/lib/perl5/site_perl/5.11.5
    /home/nick/Sandpit/snap5.9.x-v5.11.5-89-gab8e66c/lib/perl5/5.11.5/ia64-linux-ld
    /home/nick/Sandpit/snap5.9.x-v5.11.5-89-gab8e66c/lib/perl5/5.11.5
    .


Environment for perl 5.11.5:
    HOME=/home/nick
    LANG=en_US.UTF-8
    LANGUAGE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Mar 10\, 2010 at 12&#8203;:00&#8203;:53PM \-0800\, Nicholas Clark wrote&#8203;:

> This is a build failure report for perl from nick@&#8203;ccl4\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.11\.5\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> \.\./lib/overload\.t \.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\. 1/607
> \#   Failed test 'numifies to usual reference value'
> \#   at \.\./lib/overload\.t line 1561\.
> \#          got&#8203;: '6\.91752902764528267e\+18'
> \#     expected&#8203;: '6917529027645282672'
> 
> \#   Failed test 'division of ref'
> \#   at \.\./lib/overload\.t line 1572\.
> \#          got&#8203;: '6\.91752902764528267e\+18'
> \#     expected&#8203;: '6917529027645282672'
> \# Looks like you failed 2 tests of 607\.
> \.\./lib/overload\.t \.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\. Dubious\, test returned 2 \(wstat 512\, 0x200\)
> Failed 2/607 subtests
> 
> The tests in question are
> 
>     is\($m/1\, $num\_val\, 'numifies to usual reference value'\);
> 
>     is\($aref/1\, $num\_val\, 'division of ref'\);
> 
> where $m and $aref are created like this&#8203;:
> 
>     package numify\_by\_fallback;
>     use overload fallback => 1;
> 
>     package main;
> 
>     my $aref = \[\];
>     my $m = bless $aref\, 'numify\_by\_fallback';
> 
> 
> The test is making unwarrented assumptions about the numeric values of the
> memory map\. On ia64 Linux the values are BIG&#8203;:
> 
> $ \./perl \-e 'printf "%x\\n"\, \[\]'
> 600000000001f0e0
> 
> I'm not sure what the right fix is\. I don't think that this constitutes a
> blocker\.

As it's only a test that is in error\, fixed with&#8203;:

commit 9fa8ecf2a3ca91ec626e605a91e2840b8755036e
Author&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
Date&#8203;:   Thu Mar 11 14&#8203;:08&#8203;:07 2010 \+0000

&nbsp;    Fix stringification assumption bug in overload\.t\, revealed by ia64\-linux\-ld\.
&nbsp;    
&nbsp;    Specifically&#8203;:
&nbsp;    1&#8203;: / returns and NV where possible\, only returning an integer if the dividend
&nbsp;       is an integer larger than an NV can represent accurately\, and integer
&nbsp;       division is exact \(ie no fractional part/remainder\)\.
&nbsp;    2&#8203;: The test is performing $ref/1\, intending it to be an identity operation
&nbsp;       on the numeric value of the reference\.
&nbsp;    3&#8203;: The test assumes that the return result of the division will be a number
&nbsp;       that stringifies identically to the integer value of the reference\.
&nbsp;    
&nbsp;    The fails if both&#8203;:
&nbsp;    
&nbsp;    1&#8203;: The system memory map is such that addresses are very large \(ia64 does\)
&nbsp;    2&#8203;: NVs are large enough to hold these addresses
&nbsp;    
&nbsp;    because then the address becomes converted to an NV which has sufficient
&nbsp;    decimal digits that stringification defaults to scientific notation\.
&nbsp;    
&nbsp;    Itanium Linux users the world over will be cheering because they can now
&nbsp;    compile Perl with long doubles with confidence that all tests pass\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/lib/overload.t b/lib/overload.t
index 39333cf..734e8b1 100644
--- a/lib/overload.t
+++ b/lib/overload.t
@@ -1558,7 +1558,7 @@ foreach my $op (qw(<=> == != < <= > >=)) {
     is($m+$m, 2*$num_val, 'numifies to usual reference value');
     is(0-$m, -$num_val, 'numifies to usual reference value');
     is(1*$m, $num_val, 'numifies to usual reference value');
-    is($m/1, $num_val, 'numifies to usual reference value');
+    is(int($m/1), $num_val, 'numifies to usual reference value');
     is($m%100, $num_val%100, 'numifies to usual reference value');
     is($m**1, $num_val, 'numifies to usual reference value');
 
@@ -1569,7 +1569,7 @@ foreach my $op (qw(<=> == != < <= > >=)) {
     is($aref+$aref, 2*$num_val, 'ref addition');
     is(0-$aref, -$num_val, 'subtraction of ref');
     is(1*$aref, $num_val, 'multiplicaton of ref');
-    is($aref/1, $num_val, 'division of ref');
+    is(int($aref/1), $num_val, 'division of ref');
     is($aref%100, $num_val%100, 'modulo of ref');
     is($aref**1, $num_val, 'exponentiation of ref');
 }

Nicholas Clark
```
</details>

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'new' to 'resolved'
